src/packages/utils/src/things/logger.ts

```diff
@@ -0,0 +1,61 @@
+import { appendFile } from "fs";
+import { promisify, format } from "util";
+const appendFilePromise = promisify(appendFile);
+export type LogFunc = (message?: any, ...optionalParams: any[]) => void;
+
+type SyncronousLogger = {
+  log: LogFunc;
+};
+type AsyncronousLogger = SyncronousLogger & {
+  getCompletionHandle: () => Promise<void>;
+};
+
+export type Logger = SyncronousLogger | AsyncronousLogger;
+
+export function createLogger(config: {
+  file: number;
+  baseLogger: Logger;
+}): AsyncronousLogger;
+export function createLogger(config: { baseLogger: Logger }): SyncronousLogger;
+export function createLogger(config: {
+  file?: number;
+  baseLogger: Logger;
+}): Logger {
+  if (config.file === undefined) {
+    return config.baseLogger;
+  } else {
+    if (typeof config.file !== "number") {
+      throw new Error(
+        `'config.file' was not correctly noramlized to a file descriptor. This should not happen. ${
+          config.file
+        }: ${typeof config.file}`
+      );
+    }
+    const descriptor = config.file;
+
+    const diskLogFormatter = (message: any) => {
+      const linePrefix = `${new Date().toISOString()} `;
```

Comment:
Probably worth documenting if that trailing space is necessary so a future maintainer won't trim it.

Reply:
Done